File: Src/Engine/Runtime/UI/GUI/Panels/SplitPanel.cpp

#include "SplitPanel.h"

#include "Panel.h"
#include "Runtime/Render/2D/Render2D.h"
#include "Runtime/UI/GUI/Style.h"

namespace SE
{
	void SplitPanel::UpdateSplitRect()
	{
		if (_orientation == Orientation::Horizontal)
		{
			float split = Math::RoundToInt(_splitterValue * Width);
			_splitterRect = Rectangle(Math::Clamp(split - SplitterSizeHalf, 0.0f, Width.operator->()), 0, SplitterSize, Height);
		}
		else
		{
			float split = Math::RoundToInt(_splitterValue * Height);
			_splitterRect = Rectangle(0, Math::Clamp(split - SplitterSizeHalf, 0.0f, Height.operator->()), Width, SplitterSize);
		}
	}

	void SplitPanel::StartTracking()
	{
		// Start move
		_splitterClicked = true;

		// Start capturing mouse
		StartMouseCapture();
	}

	void SplitPanel::EndTracking()
	{
		if (_splitterClicked)
		{
			// Clear flag
			_splitterClicked = false;

			// End capturing mouse
			EndMouseCapture();
		}
	}

	SplitPanel::SplitPanel(::SE::Orientation orientation, ScrollBars panel1Scroll, ScrollBars panel2Scroll)
	{
		AutoFocus = false;

		_orientation = orientation;
		_splitterValue = 0.5f;

		Panel1 = New<Panel>(panel1Scroll);
		Panel2 = New<Panel>(panel2Scroll);

		Panel1->Parent = this;
		Panel2->Parent = this;

		UpdateSplitRect();
	}

	void SplitPanel::Draw()
	{
		ContainerControl::Draw();

		// Draw splitter
		Style* style = Style::Current;
		Render2D::FillRectangle(_splitterRect, _splitterClicked ? style->BackgroundSelected : _mouseOverSplitter ? style->BackgroundHighlighted : style->LightBackground);
	}

	void SplitPanel::OnLostFocus()
	{
		EndTracking();

		ContainerControl::OnLostFocus();
	}

	void SplitPanel::OnMouseMove(Float2 location)
	{
		_mouseOverSplitter = _splitterRect.Contains(location);

		if (_splitterClicked)
		{
			SplitterValue = _orientation == Orientation::Horizontal ? location.x / Width : location.y / Height;
			Cursor = _orientation == Orientation::Horizontal ? CursorType::SizeWE : CursorType::SizeNS;
			_cursorChanged = true;
		}
		else if (_mouseOverSplitter)
		{
			Cursor = _orientation == Orientation::Horizontal ? CursorType::SizeWE : CursorType::SizeNS;
			_cursorChanged = true;
		}
		else if (_cursorChanged)
		{
			Cursor = CursorType::Default;
			_cursorChanged = false;
		}

		ContainerControl::OnMouseMove(location);
	}

	bool SplitPanel::OnMouseDown(Float2 location, MouseButton button)
	{
		if (button == MouseButton::Left && _splitterRect.Contains(location))
		{
			// Start moving splitter
			StartTracking();
			Focus();
			return true;
		}

		return ContainerControl::OnMouseDown(location, button);
	}

	bool SplitPanel::OnMouseUp(Float2 location, MouseButton button)
	{
		if (_splitterClicked)
		{
			EndTracking();
			return true;
		}

		return ContainerControl::OnMouseUp(location, button);
	}

	void SplitPanel::OnMouseLeave()
	{
		// Clear flag
		_mouseOverSplitter = false;
		if (_cursorChanged)
		{
			Cursor = CursorType::Default;
			_cursorChanged = false;
		}

		ContainerControl::OnMouseLeave();
	}

	void SplitPanel::OnEndMouseCapture()
	{
		EndTracking();
	}

	void SplitPanel::OnSizeChanged()
	{
		ContainerControl::OnSizeChanged();

		UpdateSplitRect();
		PerformLayout();
	}

	void SplitPanel::PerformLayoutBeforeChildren()
	{
		ContainerControl::PerformLayoutBeforeChildren();

		if (_orientation == Orientation::Horizontal)
		{
			float split = Math::RoundToInt(_splitterValue * Width);
			Panel1->Bounds = Rectangle(0, 0, split - SplitterSizeHalf, Height);
			Panel2->Bounds = Rectangle(split + SplitterSizeHalf, 0, Width - split - SplitterSizeHalf, Height);
		}
		else
		{
			float split = Math::RoundToInt(_splitterValue * Height);
			Panel1->Bounds = Rectangle(0, 0, Width, split - SplitterSizeHalf);
			Panel2->Bounds = Rectangle(0, split + SplitterSizeHalf, Width, Height - split - SplitterSizeHalf);
		}
	}

	void SplitPanel::__SetOrientation(::SE::Orientation value)
	{
		if (_orientation != value)
		{
			_orientation = value;
			UpdateSplitRect();
			PerformLayout();
		}
	}

	void SplitPanel::__SetSplitterValue(float value)
	{
		value = Math::Saturate(value);
		if (!Math::IsNearEqual(_splitterValue, value))
		{
			// Set new value
			_splitterValue = value;

			// Calculate rectangle and update panels
			UpdateSplitRect();
			PerformLayout();
		}
	}
} // SE

File: Src/Engine/Runtime/UI/GUI/Panels/SplitPanel.cs
using System;

namespace SE.GUI
{
    /// <summary>
    /// Contains two panels separated by a draggable splitter.
    /// </summary>
    public class SplitPanel : ContainerControl
    {
        public const int SplitterSize = 4;

        private Orientation m_Orientation;
        private float m_SplitterValue = 0.5f;
        private Rectangle m_SplitterBounds;
        private bool m_IsTrackingSplitter;
        private bool m_IsMouseOverSplitter;

        public SplitPanel(Orientation orientation = Orientation.Horizontal)
        {
            AutoFocus = false;
            m_Orientation = orientation;
            Panel1 = new Panel();
            Panel2 = new Panel();
            AddChild(Panel1);
            AddChild(Panel2);
            UpdateSplitterBounds();
        }

        public Panel Panel1 { get; }
        public Panel Panel2 { get; }

        public Orientation Orientation
        {
            get => m_Orientation;
            set
            {
                if (m_Orientation == value)
                    return;
                m_Orientation = value;
                PerformLayout();
            }
        }

        public float SplitterValue
        {
            get => m_SplitterValue;
            set
            {
                float clamped = Math.Clamp(value, 0.0f, 1.0f);
                if (MathF.Abs(m_SplitterValue - clamped) <= float.Epsilon)
                    return;
                m_SplitterValue = clamped;
                PerformLayout();
            }
        }

        public Rectangle SplitterBounds => m_SplitterBounds;

        public override bool OnMouseDown(Float2 location, MouseButton button)
        {
            if (button != MouseButton.Left || !m_SplitterBounds.Contains(location))
                return false;

            m_IsTrackingSplitter = true;
            Root?.StartTrackingMouse(this);
            return true;
        }

        public override void OnMouseMove(Float2 location)
        {
            m_IsMouseOverSplitter = m_SplitterBounds.Contains(location);
            if (m_IsTrackingSplitter)
                SplitterValue = m_Orientation == Orientation.Horizontal ? location.X / MathF.Max(1.0f, Width) : location.Y / MathF.Max(1.0f, Height);
        }

        public override bool OnMouseUp(Float2 location, MouseButton button)
        {
            if (button != MouseButton.Left || !m_IsTrackingSplitter)
                return false;

            m_IsTrackingSplitter = false;
            Root?.EndTrackingMouse();
            return true;
        }

        public override void ClearState()
        {
            m_IsTrackingSplitter = false;
            m_IsMouseOverSplitter = false;
            base.ClearState();
        }

        public override void Draw()
        {
            base.Draw();
            if (!VisibleInHierarchy || IsDisposed)
                return;

            Rectangle splitter = new Rectangle(ScreenPos + m_SplitterBounds.Location, m_SplitterBounds.Size);
            Color color = m_IsTrackingSplitter ? Style.Current.BackgroundSelected : m_IsMouseOverSplitter ? Style.Current.BackgroundHighlighted : Style.Current.BackgroundNormal;
            Render2D.FillRectangle(ref splitter, ref color);
        }

        protected override void OnLayoutChildren()
        {
            UpdateSplitterBounds();
            float half = SplitterSize * 0.5f;
            if (m_Orientation == Orientation.Horizontal)
            {
                float split = MathF.Round(m_SplitterValue * Width);
                Panel1.SetBounds(0.0f, 0.0f, MathF.Max(0.0f, split - half), Height);
                Panel2.SetBounds(MathF.Min(Width, split + half), 0.0f, MathF.Max(0.0f, Width - split - half), Height);
            }
            else
            {
                float split = MathF.Round(m_SplitterValue * Height);
                Panel1.SetBounds(0.0f, 0.0f, Width, MathF.Max(0.0f, split - half));
                Panel2.SetBounds(0.0f, MathF.Min(Height, split + half), Width, MathF.Max(0.0f, Height - split - half));
            }
        }

        private void UpdateSplitterBounds()
        {
            if (m_Orientation == Orientation.Horizontal)
            {
                float split = MathF.Round(m_SplitterValue * Width);
                m_SplitterBounds = new Rectangle(Math.Clamp(split - SplitterSize * 0.5f, 0.0f, Width), 0.0f, SplitterSize, Height);
            }
            else
            {
                float split = MathF.Round(m_SplitterValue * Height);
                m_SplitterBounds = new Rectangle(0.0f, Math.Clamp(split - SplitterSize * 0.5f, 0.0f, Height), Width, SplitterSize);
            }
        }
    }
}
